### PR TITLE
tests: Add ngclient root key rotation tests

### LINKED ISSUE #1607: tests: Updater key rotation tests
This is part of #1579, we want to test the Updater in various key/signature change scenarios in a setup where:
* tests are readable / understandable
* a new test scenario adds as little code as possible
* adding new scenarios should be easy and the added test runtime should be negligible

I have been looking at existing tests (_test_key_revocation_integration.py,  test_root_versioning_integration.py and test_updater_root_rotation_integration.py_) and I believe they all revolve around a few variables that we should be able to turn into a table testing setup where no additional code is needed for most new tests. The most interesting and complex case is root rotations (because root delegates not just to other roles but to "future root role") so I'll start there.

All root key/signature change cases can be defined by listing N sequential root states, each with
* key list
* threshold
* signature list
* expected result

In practice a table like this should do it:

```python
@dataclass
class RootVersion:
    keys: list[int]
    threshold: int
    signatures: list[int]
    result: Optional[Exception] = None

root_cases = {
    "key rotation: 1-of-1 scenario" : [
        RootVersion([1], 1, [1]),
        RootVersion([2], 1, [2,1]),
        RootVersion([2], 1, [2]),
    ],
    "failing threshold decrease: 2-of-2 becomes 1-of-2" : [
        RootVersion([1,2], 2, [1,2]),
        RootVersion([1,2], 1, [2], UnsignedMetadataError),
    ]
}
```
The "key rotation" test case written out would be
* root v1 contains root key1 with threshold of 1, signed by key1
* root v2 contains root key2 with threshold of 1, signed by key1 and key2
* root v2 contains root key2 with threshold of 1, signed by key2

There is no exception listed so Updater is expected to load all the versions correctly

The "failing" test case written out would be
* root v1 contains root key1 and key2 with threshold of 2, signed by key1 and key2
* root v2 contains root key1 and key2 with threshold of 1, signed by key2

Updater should fail to update v2 as it's not signed by both keys as v1 requires

This is not super readable (using numbers to represent keys is admittedly a bit confusing) but similar 3-5 lines can describe quite tricky situations like threshold increases while rotating keys etc

### PR DESCRIPTION
**Pull request by jku**
Add 15 key rotation cases to ngclient tests.

Maybe fixes #1607? This only tests root keys but I think this is the tricky part: maybe there should be a small test for other keys as well...

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: tests/repository_simulator.py [hunk 1]
```diff
@@ -134,10 +135,16 @@ def all_targets(self) -> Iterator[Tuple[str, Targets]]:
...
-    def create_key(self) -> Tuple[Key, SSlibSigner]:
+    @staticmethod
+    def create_key() -> Tuple[Key, SSlibSigner]:
         sslib_key = generate_ed25519_key()
         return Key.from_securesystemslib_key(sslib_key), SSlibSigner(sslib_key)
 
+    def add_signer(self, role:str, signer: SSlibSigner):
+            if role not in self.signers:
+                self.signers[role] = {}
+            self.signers[role][signer.key_dict["keyid"]] = signer
```
Comment (jku) [inline tests/repository_simulator.py]: oops, indent issue here

**Comment by jku** [comment]
I should also fix some linter issues if we're going to enable the linter on tests...

File: tests/test_updater_key_rotations.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python
+
+# Copyright 2021, New York University and the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Test ngclient Updater using the repository simulator
```
Comment (sechkova) [inline tests/test_updater_key_rotations.py]: nit: the docstring needs to be updated.

File: tests/test_updater_key_rotations.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            RootVersion([0,1,2,3,4], 3, [0,2,4]),
+            RootVersion([0,1,2,3,4], 3, [0,4,1]),
+            RootVersion([0,1,2,3,4], 3, [0,1,3]),
+            RootVersion([0,1,2,3,4], 3, [0,1,3]),
+        ],
+        "3-of-5, one key rotated" : [
+            RootVersion([0,1,2,3,4], 3, [0,2,4]),
+            RootVersion([0,1,3,4,5], 3, [0,4,1]),
+        ],
+        "3-of-5, one key rotate fails: not signed with 3 keys from last root" : [
```
Comment (sechkova) [inline tests/test_updater_key_rotations.py]: nit: "from new root"

**Review by sechkova — COMMENTED** [review]
It looks ok to me, I didn't hit the approve button so that you fix the indent and the commits creator :)

**Comment by sechkova** [comment]
ah and yeah, I think we need to add smaller tests for other keys as well, not necessarily with this PR.

**Review by sechkova — APPROVED** [review]

**Comment by jku** [comment]
The last push is just rebase on recent RepositorySimulator changes.

> I should also fix some linter issues if we're going to enable the linter on tests...

Not promising it's 100% there (hard to test without merging with martins PR), but most of the issues should be handled: I'll merge this with the current approval.  @MVrachev if there's something exceptionally annoying when you try rebasing, let me know

**Pull request merged**